solidity/contracts/middleware/liquidity-layer/adapters/CctpAdapter.sol

```diff
@@ -0,0 +1,185 @@
+pragma solidity ^0.8.13;
+
+import {Router} from "../../../Router.sol";
+
+import {ITokenMessenger} from "../interfaces/circle/ITokenMessenger.sol";
+import {ILiquidityLayerAdapterV2} from "../interfaces/ILiquidityLayerAdapterV2.sol";
+
+import {TypeCasts} from "../../../libs/TypeCasts.sol";
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+contract CctpAdapter is ILiquidityLayerAdapterV2, Router {
+    using SafeERC20 for IERC20;
+
+    /**
+     * @dev Emitted on `transferRemote` when a transfer message is dispatched.
+     * @param destination The identifier of the destination chain.
+     * @param recipient The address of the recipient on the destination chain.
+     * @param amount The amount of tokens burnt on the origin chain.
+     */
+
+    event SentTransferRemote(
+        uint32 indexed destination,
+        bytes32 indexed recipient,
+        uint256 amount
+    );
+
+    /**
+     * @dev Emitted when the amount of gas required to process a CCTP transfer is updated.
+     * @param oldGasAmount The old gas amount.
+     * @param newGasAmount The new gas amount.
+     */
+    event GasAmountSet(
+        uint256 indexed oldGasAmount,
+        uint256 indexed newGasAmount
+    );
+
+    /// @notice The TokenMessenger contract.
+    ITokenMessenger public tokenMessenger;
+
+    /// @notice The token address.
+    address public token;
+
+    /// @notice The token symbol of the token.
+    string public tokenSymbol;
+
+    /// @notice The amount of gas required to process a CCTP transfer.
+    uint256 public gasAmount;
+
+    /// @notice Hyperlane domain => Circle domain.
+    /// ATM, known Circle domains are Ethereum = 0 and Avalanche = 1.
+    /// Note this could result in ambiguity between the Circle domain being
+    /// Ethereum or unknown.
+    mapping(uint32 => uint32) public hyperlaneDomainToCircleDomain;
+
+    /**
+     * @notice Emits the nonce of the Circle message when a token is bridged.
+     * @param nonce The nonce of the Circle message.
+     */
+    event BridgedToken(uint64 nonce);
+
+    /**
+     * @notice Emitted when the Hyperlane domain to Circle domain mapping is updated.
+     * @param hyperlaneDomain The Hyperlane domain.
+     * @param circleDomain The Circle domain.
+     */
+    event DomainAdded(uint32 indexed hyperlaneDomain, uint32 circleDomain);
+
+    /**
+     * @param _owner The new owner.
+     * @param _tokenMessenger The TokenMessenger contract.
+     * @param _token The token address.
+     * @param _tokenSymbol The token symbol.
+     * @param _gasAmount The amount of gas required to process a CCTP transfer.
+     * @param _mailbox The address of the mailbox contract.
+     * @param _interchainGasPaymaster The address of the interchain gas paymaster contract.
+     * @param _interchainSecurityModule The address of the interchain security module contract.
```

Comment:
Looking at this, I'm actually thinking whether the adapter and ISM should be the same contract on every chain? And then if that is the case, could all this just live in https://github.com/Accio-labs/permissionless-cctp

Reply:
I like this conceptually for the monorepo PR to just have the `ILiquidityLayerAdapterV2` interface as well as making `TokenRouter` inherit from it